### PR TITLE
Check clone

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
This adds checks for the clone macro compilation errors and also fixes https://github.com/gtk-rs/glib/issues/554#issuecomment-569218290

### CODE REVIEW

**Comment by sdroege** [comment]
I think you want https://crates.io/crates/trybuild :)

**Comment by GuillaumeGomez** [comment]
Didn't know this one. But considering we are in a simple case, not sure if it's worth it. If we had proc-macros with more advanced errors, it'd make sense though.

**Comment by sdroege** [comment]
It makes sense because you don't have to add a 150 line Python script for doing what that crate already does :)

**Comment by GuillaumeGomez** [comment]
150 lines seem very little to me but I see your point. What do you think @EPashkin ?

**Comment by ricvelozo** [comment]
@GuillaumeGomez Would it be possible to support closures without curly braces, like common Rust? Specially for simpler cases.

**Comment by GuillaumeGomez** [comment]
We could. But not sure if we want it.

@EPashkin @sdroege: opinion?

**Comment by sdroege** [comment]
Sure, why not? More consistent with how closures work otherwise.

**Comment by GuillaumeGomez** [comment]
Updatd. Tried trybuild, needs to be improved before we can use it here.

**Comment by sdroege** [comment]
> Updatd. Tried trybuild, needs to be improved before we can use it here.

Can you create an issue against trybuild with your problems so it is not forgotten? :)

**Comment by GuillaumeGomez** [comment]
I did better: I created a calendar event so I get spammed every day until I've done it.

**Comment by sdroege** [comment]
ok :) Anything missing here? Do you also want this in the 0.9.1 bugfix release?

File: src/clone.rs [hunk 1]
```diff
@@ -389,9 +410,24 @@ macro_rules! clone {
...
+                let return_value = || $crate::to_return_value!($($return_value)?);
+                $( $crate::to_type_after!($(as $rename)? $(@ $strength)? $($variables).+, return_value);)*
+                $body
+            }
+        }
+    );
+    ($($(@ $strength:ident)? $($variables:ident).+ $(as $rename:ident)?),+ => $(@default-return $return_value:expr,)? move | $($pattern:pat),* | $($body:tt)* ) => (
+        // In case we have:
+        // clone!(@weak foo => move |bla| bla);
+        compile_error!("Closure body needs to be surrounded with curly braces");
```
Comment (sdroege) [inline src/clone.rs]: I wonder if rustfmt would remove the curly braces if the closure body is too small. What does it do, can you check? :)

If rustfmt removes them then we need to support that case unfortunately.

Comment (GuillaumeGomez) [reply src/clone.rs]: Normally it's supposed to be removed. Forgot to remove this one I guess...

Comment (sdroege) [reply src/clone.rs]: I mean a test that succeeds to compile and run :)

Comment (GuillaumeGomez) [reply src/clone.rs]: I updated one in the doc. :)

File: src/clone.rs [hunk 2]
```diff
@@ -360,6 +362,15 @@ macro_rules! clone {
             }
         }
     );
+    ($($(@ $strength:ident)? $($variables:ident).+ $(as $rename:ident)?),+ => @default-panic, move || $body:expr ) => (
```
Comment (sdroege) [inline src/clone.rs]: Couldn't this just call itself and put `{}` around `$body`? Less code duplication then

File: src/clone.rs [hunk 3]
```diff
@@ -370,6 +381,16 @@ macro_rules! clone {
             }
         }
     );
+    ($($(@ $strength:ident)? $($variables:ident).+ $(as $rename:ident)?),+ => $(@default-return $return_value:expr,)? move || $body:expr ) => (
```
Comment (sdroege) [inline src/clone.rs]: Same here

File: src/clone.rs [hunk 4]
```diff
@@ -389,9 +410,19 @@ macro_rules! clone {
             }
         }
     );
+    ($($(@ $strength:ident)? $($variables:ident).+ $(as $rename:ident)?),+ => @default-panic, move | $($pattern:pat),* | $body:expr ) => (
```
Comment (sdroege) [inline src/clone.rs]: And here

**Comment by GuillaumeGomez** [comment]
Added more tests for the macro.

File: src/clone.rs [hunk 5]
```diff
@@ -360,16 +356,22 @@ macro_rules! clone {
...
+    ($($(@ $strength:ident)? $($variables:ident).+ $(as $rename:ident)?),+ => @default-panic, move || $body:expr ) => (
+        clone!($($(@ $strength)? $($variables).+ $(as $rename)?),* => @default-panic, move || { $body })
+    );
     ($($(@ $strength:ident)? $($variables:ident).+ $(as $rename:ident)?),+ => $(@default-return $return_value:expr,)? move || $body:block ) => (
         {
             $( $crate::to_type_before!($(@ $strength)? $($variables).+ $(as $rename)?); )*
             move || {
-                let return_value = || $crate::to_return_value!($($return_value)?);
-                $( $crate::to_type_after!($(as $rename)? $(@ $strength)? $($variables).+, return_value );)*
+                let _return_value = || $crate::to_return_value!($($return_value)?);
```
Comment (sdroege) [inline src/clone.rs]: Why the underscore?

**Comment by sdroege** [comment]
Looks good to me otherwise

File: src/clone.rs (hunk 5 above)
Comment (GuillaumeGomez) [reply src/clone.rs]: In case you don't use the returned value, it avoids getting a warning about it (discovered it when adding the tests).

Comment (sdroege) [reply src/clone.rs]: Ah, makes sense. Merge when the CI is green please, or I'll merge it later once I notice that the CI has passed :)

I'll also do a backport for 0.9.1 then and then we can release, or what do you think?

Comment (GuillaumeGomez) [reply src/clone.rs]: Sounds good to me!

**Comment by GuillaumeGomez** [comment]
Fixed typos and forgot to update the tests. :)

**Comment by GuillaumeGomez** [comment]
And now formatting...

**Comment by sdroege** [comment]
Hmm that formatting looks like rustfmt can't figure out our macro correctly. I think you'll want to format this like following, or similar. Why do we have the arrow in the macro there btw instead of just a comma?

```rust
        let _ = clone!(
            @strong v,
            @strong w as x =>
            @default-panic,
            move || println!("v: {}, w: {}", v, x)
        );
```

**Comment by GuillaumeGomez** [comment]
> Why do we have the arrow in the macro there btw instead of just a comma?

To make the reading easier by splitting the arguments and the rest visually.

**Comment by sdroege** [comment]
@GuillaumeGomez You do the 0.9.1 backport and then the release?

**Comment by GuillaumeGomez** [comment]
Sure.

**Pull request merged**